src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -145,164 +160,115 @@ public void Shutdown()
-                    {
-                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can react
-                        completion.SetException(exception);
-
-                        // Also allow the exception to flow back to the request queue to handle as appropriate
-                        throw new InvalidOperationException($"Error handling '{methodName}' request: {exception.Message}", exception);
-                    }
-                }, requestCancellationToken);
+                combinedCancellationToken);
 
-            var didEnqueue = _queue.TryEnqueue(item);
+            var didEnqueue = _queue.TryEnqueue((item, combinedCancellationToken));
 
             // If the queue has been shut down the enqueue will fail, so we just fault the task immediately.
             // The queue itself is threadsafe (_queue.TryEnqueue and _queue.Complete use the same lock).
             if (!didEnqueue)
             {
-                completion.SetException(new InvalidOperationException($"{_serverName} was requested to shut down."));
+                return Task.FromException<TResponseType?>(new InvalidOperationException($"{_serverName} was requested to shut down."));
             }
 
-            return completion.Task;
+            return resultTask;
         }
 
         private async Task ProcessQueueAsync()
         {
-            QueueItem? inProgressWorkItem = null;
             try
             {
                 while (!_cancelSource.IsCancellationRequested)
                 {
-                    var work = await _queue.DequeueAsync(_cancelSource.Token).ConfigureAwait(false);
-                    inProgressWorkItem = work;
-
-                    // Record when the work item was been de-queued and the request context preparation started.
-                    work.Metrics.RecordExecutionStart();
 
-                    // Restore our activity id so that logging/tracking works across asynchronous calls.
-                    Trace.CorrelationManager.ActivityId = work.ActivityId;
-                    var context = CreateRequestContext(work);
+                    // First attempt to de-queue the work item in its own try-catch.
+                    // This is because before we de-queue we do not have access to the queue item's linked cancellation token.
+                    (IQueueItem work, CancellationToken cancellationToken) queueItem;
+                    try
+                    {
+                        queueItem = await _queue.DequeueAsync(_cancelSource.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException ex) when (ex.CancellationToken == _cancelSource.Token)
+                    {
+                        // The queue's cancellation token was invoked which means we are shutting down the queue.
+                        // Exit out of the loop so we stop processing new items.
+                        return;
+                    }
```

Comment:
note: if you don't like this try/catch, i believe @genlu is adding an extension to await this without throign cancellation.  You can then just await, and tehn do a quick check/bail if the _cancelSource is canceled.